Verify if lvalue is in known arguments

yamatoRT opened the pull request:
Hello, 
This PR solve this issue : [#3047](https://github.com/robotframework/robotframework/issues/3047)
It's my first PR, don't hesitate to give me advices.

Regards,

pekkaklarck commented:
Big thanks for the PR and sorry for review taking so long. I'm now concentrating on RF 3.1.2 and try to get it release at the end of April.

The fix itself looks very good. The only comment I have about the code is that the added empty line should be removed.

Tests also look good. We do, however, already have tests for named-only-arguments and the new test should be added there. The functionality ought to be pretty well tested so new tests are only needed for this particular bug. There are existing tests both for [user keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/user_keyword.robot) and for [library keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/python.robot) and I guess tests could be added to both.

yamatoRT commented:
Hello,
thanks for you response. I moved my tests in [user keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/user_keyword.robot) and [library keywords](https://github.com/robotframework/robotframework/blob/master/atest/robot/keywords/named_only_args/python.robot). My modification create regressions, because i had this error after i was running test : 
```
 Unexpected keyword argumemt:
Expected: Keyword 'Kw Only Arg' got unexpected named argument 'invalid'.
Actual: Keyword 'Kw Only Arg' got positional argument after named arguments.
```
Then, i prefer consider all arguments in format 'name=value' after the first named argument as named arguments and not as positional arguments.

Moreover, i removed two typos in name of test case (argumemt become argument).

I'm waiting for your feedback.

Regards,

pekkaklarck commented:
Two comments about the PR:

1. Most of the added tests are duplicate to existing tests. Only tests related to the actual change are needed. If there are missing tests, they can be added as a separate PR. Small changes like typo fixes are fine.

2. You are right that the original simple fix made error messages worse with cases like `nonex=value`. The added check ought to fix that . It also changes error messages in some other cases and related tests would need to be updated.

I've started to think is fixing this a good idea in general. It is somewhat annoying that you cannot nowadays should argument `foo=bar` be evaluated as a positional argument with literal value `foo=bar` or does it mean name argument `foo` with value `bar`. There have been some discussions about changing the syntax so that `foo=bar` would *always* be considered named argument syntax and literal value would require escaping like `foo\=bar`. That would be a big backwards incompatible change requiring long deprecation period and it may not be worth it. If we ever consider doing that, now allowing `foo=bar` as a literal value with named-only-arguments doesn't make much sense.

Let me think about this a bit more before continuing with this PR.

pekkaklarck commented:
OK, let's do this. At the moment named-only arguments behave inconsistently compared to other arguments and that's not a good thing.

The current implementation is good, but I got some style related comments about the code. I'd also like tests to be enhanced. That basically means removing most of the added tests as they are duplicate to tests that already exist. I'll add more detailed review comments about these.

yamatoRT commented on atest/testdata/keywords/named_only_args/KwOnlyArgs.py:
```diff
@@ -31,3 +31,7 @@ def all_arg_types(pos_req, pos_def='pd', *varargs,
+
+
+def kw_two_args_with_varargs(*varargs, kwo='default', another='another'):
+    return '-'.join(varargs + (kwo, another,))
```
I removed keywork. Now, I use  `Kw Only Arg ` and `Kw Only Arg With Varargs`.

yamatoRT commented on atest/testdata/keywords/named_only_args/python.robot:
```diff
@@ -44,14 +44,22 @@ Missing multiple values
+Multiple unexpected keyword argument with escaped inequality
+    [Documentation]    FAIL Keyword 'KwOnlyArgs.Kw Only Arg' got positional argument after named arguments.
+    Kw Only Arg    kwo=value    <\=ooops    ooops=invalid
+
```
I removed test case.

yamatoRT commented on atest/testdata/keywords/named_only_args/python.robot:
```diff
@@ -75,3 +83,19 @@ With other arguments
+    Kw Two Args With Varargs    First arg    Second arg    Maybe a third    kwo=Actually, correct!
+
+With only one named argument but we have a escaped egal
+    Kw Two Args With Varargs    Let's try this again    1    <\=    2    kwo=Correct again!
```
I removed test case.

yamatoRT commented on atest/testdata/keywords/named_only_args/python.robot:
```diff
@@ -75,3 +83,19 @@ With other arguments
+    Kw Two Args With Varargs    One more time    1    <=    2    kwo=This part fails
+
+With only one named argument but we have two no escaped egals
+    Kw Two Args With Varargs    One more time    a=1    <=    2    kwo=This part fails
```
I renamed test case. I changed value of kwo argument. I don't understand your last comment.

yamatoRT commented on atest/testdata/keywords/named_only_args/user_keyword.robot:
```diff
@@ -80,6 +88,23 @@ With other arguments
+With two named arguments but with a error
+    [Documentation]  FAIL    Keyword 'Kw Two Args With Varargs' got positional argument after named arguments.
+    Kw Two Args With Varargs    One more time    another=1    2      <=     kwo=This part fails
+
```
I removed test case.

yamatoRT commented on atest/testdata/keywords/named_only_args/user_keyword.robot:
```diff
@@ -115,3 +140,8 @@ All Arg Types
+Kw Two Args With Varargs
+    [Arguments]    @{varargs}    ${kwo}=default   ${another}=another
+    ${result} =    Catenate    SEPARATOR=-    @{varargs}    ${kwo}    ${another}
+    [Return]    ${result}
```
I removed keyword.

yamatoRT commented on src/robot/running/arguments/argumentresolver.py:
```diff
@@ -63,13 +63,16 @@ def _is_dict_var(self, arg):
                 VariableSplitter(arg).is_dict_variable())
 
-    def _is_named(self, arg, variables=None):
+    def _is_named(self, arg, have_even_named_argument, variables=None):
```
I named argument `previous_named`.

yamatoRT commented on src/robot/running/arguments/argumentresolver.py:
```diff
@@ -63,13 +63,16 @@ def _is_dict_var(self, arg):
         if not (is_string(arg) and '=' in arg):
             return False
         name, value = split_from_equals(arg)
+
```
I removed empty line.

yamatoRT commented on src/robot/running/arguments/argumentresolver.py:
```diff
@@ -63,13 +63,16 @@ def _is_dict_var(self, arg):
-        if self._argspec.kwargs or self._argspec.kwonlyargs:
+        if self._argspec.kwargs or name in self._argspec.kwonlyargs:
+            return True
+        if have_even_named_argument and value is not None:
```
Yes, my bad. I removed it.

yamatoRT commented on atest/testdata/keywords/named_only_args/python.robot:
```diff
@@ -75,3 +83,19 @@ With other arguments
+
+With two named arguments but with a error
+    [Documentation]  FAIL    Keyword 'KwOnlyArgs.Kw Two Args With Varargs' got positional argument after named arguments.
+    Kw Two Args With Varargs    One more time    another=1    2      <=     kwo=This part fails
```
Yes, test already exists. I removed it.

yamatoRT commented on atest/testdata/keywords/named_only_args/user_keyword.robot:
```diff
@@ -48,14 +48,22 @@ Missing multiple values
+Multiple unexpected keyword argument with escaped inequality
+    [Documentation]    FAIL Keyword 'Kw Only Arg' got positional argument after named arguments.
+    Kw Only Arg    kwo=value    <\=ooops    ooops=invalid
+
```
I removed test case.

pekkaklarck commented:
Thanks for the changes! There are small things that could still be enhanced in tests, but I'll merge this now and do those changes myself afterwards.

pekkaklarck commented:
Found a small bug in the implementation. It is possible to use `${var}=value` and in that case `${var}` should be resolved before checking is `name in argspec.kwonlyargs`. Resolving is already done before checking is `name in argspec.positional` and fixing the bug only requires doing that a bit earlier. I'll add tests and fix this myself.
